### PR TITLE
#32184 save total_vocab_size

### LINKED ISSUE #32184: DataCollatorForLanguageModeling is (unnecessary) slow
### System Info

- `transformers` version: 4.43.1
- Platform: Linux-3.10.0-1062.18.1.el7.x86_64-x86_64-with-glibc2.17
- Python version: 3.11.3
- Huggingface_hub version: 0.23.4
- Safetensors version: 0.4.3
- Accelerate version: 0.32.1
- Accelerate config: 	not found
- PyTorch version (GPU?): 1.13.0+cu117 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No

### Who can help?

@ArthurZucker

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```
from transformers import AutoTokenizer
from transformers import DataCollatorForLanguageModeling

tokzr = AutoTokenizer.from_pretrained('bert-base-multilingual-cased', use_fast=False)

masker = DataCollatorForLanguageModeling(tokzr)
data = ['this is a very boring test sentence'] * 1000

for line in data:
    token_ids = tokzr.encode(line, return_tensors='pt')[0]
    #input_text, output_labels = masker.torch_mask_tokens(token_ids.view(1, -1))
```

### Expected behavior

I would expect this code to be not much slower when uncommenting the masking. However, uncommenting it makes the code 5 times slower. I have tried to profile where this happens, and it seems like it is mainly happening in finding the random words to replace with, as it gets the length of the vocabulary there which takes a long time (https://github.com/huggingface/transformers/blob/main/src/transformers/data/data_collator.py#L854). I am not sure how to make getting the length faster, but perhaps it can just be saved/cached?

### PR DESCRIPTION
**Pull request by itazap**
fixes #32184 save total_vocab_size to speed up operation

- updates` len(tokenizer)` to `len(self.get_vocab())` from `len(set(self.get_vocab().keys()))` as it is equivalent for Python 2.x and 3.x due to dict key uniqueness.

- stores an internal variable called total_vocab_size to avoid bottleneck in `len(self.get_vocab()`.

TESTS:

- existing tests check `len(tokenizer)` and pass! (see: `test_tokenization_common` --> `test_add_tokens`

@ArthurZucker

### CODE REVIEW

**Comment by LysandreJik** [comment]
Leaving the review to @ArthurZucker, tokenizer mastermind :)

File: src/transformers/tokenization_utils.py [hunk 1]
```diff
@@ -494,10 +494,17 @@ def get_added_vocab(self) -> Dict[str, int]:
...
-        return len(set(self.get_vocab().keys()))
+        return self.total_vocab_size
+
+    def _update_total_vocab_size(self):
+        """
+        Update the size of the full vocabulary with the added tokens. Counts the `keys` and not the `values` because
+        otherwise if there is a hole in the vocab, we will add tokenizers at a wrong index. This operation is slow and
+        is only updated when adding tokens.
+        """
+        self.total_vocab_size = len(self.get_vocab())
```
Comment (amyeroberts) [inline src/transformers/tokenization_utils.py:507]: Given `_add_tokens` already keeps a running count of the number of tokens added, couldn't we just add these to the total rather than re-count each time?

Comment (itazap) [reply src/transformers/tokenization_utils.py]: yes agreed! Pushed this change

Comment (amyeroberts) [reply src/transformers/tokenization_utils.py]: Nice :) I guess the only thing to consider is are there methods to remove tokens too? Do we expect people to manually edit the vocab mapping?

Comment (itazap) [reply src/transformers/tokenization_utils.py]: hmm technically no, but if users play around with popping out tokens then it would not be updated. Maybe we keep  `__len__ ` dynamic to be re-counted each time then? it's not heavy in terms of run time

Comment (amyeroberts) [reply src/transformers/tokenization_utils.py]: Sounds good!

File: src/transformers/tokenization_utils.py [hunk 2]
```diff
@@ -572,7 +572,7 @@ def _add_tokens(self, new_tokens: Union[List[str], List[AddedToken]], special_to
             self._added_tokens_encoder[token.content] = token_index
             if self.verbose:
                 logger.info(f"Adding {token} to the vocabulary")
-
+            self.total_vocab_size += 1
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils.py:575]: we should only add if `token.content not in current_vocab` !

Comment (ArthurZucker) [reply src/transformers/tokenization_utils.py:575]: Also probably in the `    @added_tokens_decoder.setter
    def added_tokens_decoder(self, value: Dict[int, Union[AddedToken, str]]) -> Dict[int, AddedToken]:`

Comment (ArthurZucker) [reply src/transformers/tokenization_utils.py:575]: as we don't add the tokens but set them!

**Review by ArthurZucker — COMMENTED** [review]
Nice!

File: src/transformers/tokenization_utils.py (hunk 2 above)
Comment (itazap) [reply src/transformers/tokenization_utils.py]: I reverted to an earlier commit that addresses these better! @ArthurZucker

**Review by ArthurZucker — COMMENTED** [review]
Let's add some small tests, but also would be nice to have (folllow up PR) this in the bemchmark to make sure this critical piece of code is kept fast!

**Comment by itazap** [comment]
added a small test in `test_tokenization_utils.py` to explicitly test len(tokenizer). Other tests also implicitly test the length behaviour, such as `test_tokenization_common.TokenizerTesterMixin.test_add_tokens_tokenizer`

**Review by ArthurZucker — APPROVED** [review]
Thanks 🤗
